mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java

```diff
@@ -0,0 +1,271 @@
+   */
+  Record deserialize(Object data) {
+    return (Record) fieldDeserializer.value(data);
+  }
+
+  private Deserializer(Schema schema, ObjectInspector fieldInspector) {
+    this.fieldDeserializer = DeserializerVisitor.visit(schema, fieldInspector);
+  }
+
+  private static class DeserializerVisitor extends SchemaWithPartnerVisitor<ObjectInspector, FieldDeserializer> {
+
+    public static FieldDeserializer visit(Schema schema, ObjectInspector objectInspector) {
+      return visit(schema, new FixNameMappingObjectInspector(schema, objectInspector), new DeserializerVisitor(),
+          new PartnerObjectInspectorByNameAccessors());
+    }
+
+    @Override
+    public FieldDeserializer schema(Schema schema, ObjectInspector inspector, FieldDeserializer deserializer) {
+      return deserializer;
+    }
+
+    @Override
+    public FieldDeserializer field(NestedField field, ObjectInspector inspector, FieldDeserializer deserializer) {
+      return deserializer;
+    }
+
+    @Override
+    public FieldDeserializer primitive(PrimitiveType type, ObjectInspector inspector) {
+      switch (type.typeId()) {
+        case BOOLEAN:
+        case INTEGER:
+        case LONG:
+        case FLOAT:
+        case DOUBLE:
+        case STRING:
+          // Generic conversions where Iceberg and Hive are using the same java object
+          return o -> ((PrimitiveObjectInspector) inspector).getPrimitiveJavaObject(o);
+        case UUID:
+          // TODO: This will not work with Parquet. Parquet UUID expect byte[], others are expecting UUID
+          return o -> UUID.fromString(((StringObjectInspector) inspector).getPrimitiveJavaObject(o));
+        case DATE:
+        case TIMESTAMP:
+        case FIXED:
+        case BINARY:
+        case DECIMAL:
+          // Iceberg specific conversions
+          return o -> ((WriteObjectInspector) inspector).convert(o);
+        case TIME:
+        default:
+          throw new IllegalArgumentException("Unsupported column type: " + type);
+      }
+    }
+
+    @Override
+    public FieldDeserializer struct(StructType type, ObjectInspector inspector, List<FieldDeserializer> deserializers) {
+      return o -> {
+        if (o == null) {
+          return null;
+        }
+
+        List<Object> data = ((StructObjectInspector) inspector).getStructFieldsDataAsList(o);
+        Record result = GenericRecord.create(type);
+
+        for (int i = 0; i < deserializers.size(); i++) {
+          Object fieldValue = data.get(i);
+          if (fieldValue != null) {
+            result.set(i, deserializers.get(i).value(fieldValue));
+          } else {
+            result.set(i, null);
+          }
+        }
+
+        return result;
+      };
+    }
+
+    @Override
+    public FieldDeserializer list(ListType listTypeInfo, ObjectInspector inspector, FieldDeserializer deserializer) {
+      return o -> {
+        if (o == null) {
+          return null;
+        }
+
+        List<Object> result = new ArrayList<>();
```

Comment:
Nit: prefer `Lists.newArrayList()`.

Reply:
I remember on some other PR got a review from someone else to avoid unnecessary guava uses.
I am find with both, but I think we should stick to one or the other. Shall it be Lists.newArrayList then?

Reply:
We don't want to bring in any additional Guava classes if possible, but using the ones that are already there is a good thing. For this case, we can easily replace map class implementations with an import change later.